File: cst/Expr.java
package org.pkl.cst;

import java.util.List;
import org.jetbrains.annotations.Nullable;
import org.pkl.parser.Span;

public sealed interface Expr {

  record This(Span span) implements Expr {}

  record Outer(Span span) implements Expr {}

  record Module(Span span) implements Expr {}

  record Null(Span span) implements Expr {}

  record BoolLiteral(boolean b, Span span) implements Expr {}

  record IntLiteral(long l, Span span) implements Expr {}

  record FloatLiteral(double d, Span span) implements Expr {}

  record SingleString(String str, Span span) implements Expr {}

  record MultiString(String str, Span span) implements Expr {}

  record Throw(Expr expr, Span span) implements Expr {}

  record Trace(Expr expr, Span span) implements Expr {}

  record ImportExpr(String importStr, Span span) implements Expr {}

  record ImportGlobExpr(String importStr, Span span) implements Expr {}

  record Read(Expr expr, Span span) implements Expr {}

  record ReadGlob(Expr expr, Span span) implements Expr {}

  record ReadNull(Expr expr, Span span) implements Expr {}

  record UnqualifiedAccess(Ident ident, List<Expr> args, Span span) implements Expr {}

  record QualifiedAccess(Expr expr, Ident ident, boolean isNullable, List<Expr> args, Span span)
      implements Expr {}

  record SuperAccess(Ident ident, List<Expr> args, Span span) implements Expr {}

  record SuperSubscript(Expr arg, Span span) implements Expr {}

  record Subscript(Expr expr, Expr arg, Span span) implements Expr {}

  record If(Expr cond, Expr then, Expr els, Span span) implements Expr {}

  record Let(Parameter par, Expr bindingExpr, Expr expr, Span span) implements Expr {}

  record FunctionLiteral(List<Parameter> args, Expr expr, Span span) implements Expr {}

  record Parenthesised(Expr expr, Span span) implements Expr {}

  record New(@Nullable Type type, ObjectBody body, Span span) implements Expr {}

  record Amends(Expr expr, ObjectBody body, Span span) implements Expr {}

  record NonNull(Expr expr, Span span) implements Expr {}

  record UnaryMinus(Expr expr, Span span) implements Expr {}

  record LogicalNot(Expr expr, Span span) implements Expr {}

  record BinaryOp(Expr expr1, Expr expr2, Operation op, Span span) implements Expr {}

  default Span getSpan() {
    return switch (this) {
      case This x -> x.span;
      case Outer x -> x.span;
      case Module x -> x.span;
      case Null x -> x.span;
      case BoolLiteral x -> x.span;
      case IntLiteral x -> x.span;
      case FloatLiteral x -> x.span;
      case SingleString x -> x.span;
      case MultiString x -> x.span;
      case Throw x -> x.span;
      case Trace x -> x.span;
      case ImportExpr x -> x.span;
      case ImportGlobExpr x -> x.span;
      case Read x -> x.span;
      case ReadGlob x -> x.span;
      case ReadNull x -> x.span;
      case UnqualifiedAccess x -> x.span;
      case QualifiedAccess x -> x.span;
      case SuperAccess x -> x.span;
      case SuperSubscript x -> x.span;
      case Subscript x -> x.span;
      case NonNull x -> x.span;
      case UnaryMinus x -> x.span;
      case LogicalNot x -> x.span;
      case If x -> x.span;
      case Let x -> x.span;
      case FunctionLiteral x -> x.span;
      case Parenthesised x -> x.span;
      case New x -> x.span;
      case Amends x -> x.span;
      case BinaryOp x -> x.span;
    };
  }
}

File: pkl-lsp/src/main/kotlin/org/pkl/lsp/cst/Expr.kt
/**
 * Copyright © 2024 Apple Inc. and the Pkl project authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.pkl.lsp.cst

sealed class Expr(override val span: Span) : PklNode(span) {
  data class This(override val span: Span) : Expr(span)

  data class Outer(override val span: Span) : Expr(span)

  data class Module(override val span: Span) : Expr(span)

  data class Null(override val span: Span) : Expr(span)

  data class BooleanLiteral(val b: Boolean, override val span: Span) : Expr(span)

  data class IntLiteral(val i: Long, override val span: Span) : Expr(span)

  data class FloatLiteral(val f: Double, override val span: Span) : Expr(span)

  data class ConstantString(val s: String, override val span: Span) : Expr(span)

  data class InterpolatedString(val exprs: List<Expr>, override val span: Span) : Expr(span) {
    init {
      for (it in exprs) it.parent = this
    }
  }

  data class Throw(val expr: Expr, override val span: Span) : Expr(span) {
    init {
      expr.parent = this
    }
  }

  data class Trace(val expr: Expr, override val span: Span) : Expr(span) {
    init {
      expr.parent = this
    }
  }

  data class Import(val uri: Ident, override val span: Span) : Expr(span)

  data class ImportGlob(val uri: Ident, override val span: Span) : Expr(span)

  data class Read(val expr: Expr, override val span: Span) : Expr(span) {
    init {
      expr.parent = this
    }
  }

  data class ReadGlob(val expr: Expr, override val span: Span) : Expr(span) {
    init {
      expr.parent = this
    }
  }

  data class ReadNull(val expr: Expr, override val span: Span) : Expr(span) {
    init {
      expr.parent = this
    }
  }

  data class UnqualifiedAccess(val ident: Ident, override val span: Span) : Expr(span)

  data class UnqualifiedMethodAccess(
    val ident: Ident,
    val args: List<Expr>,
    override val span: Span
  ) : Expr(span) {
    init {
      for (it in args) it.parent = this
    }
  }

  data class QualifiedAccess(
    val expr: Expr,
    val ident: Ident,
    val isNullable: Boolean,
    override val span: Span
  ) : Expr(span) {
    init {
      expr.parent = this
    }
  }

  data class QualifiedMethodAccess(
    val expr: Expr,
    val ident: Ident,
    val isNullable: Boolean,
    val args: List<Expr>,
    override val span: Span
  ) : Expr(span) {
    init {
      expr.parent = this
      for (it in args) it.parent = this
    }
  }

  data class SuperAccess(val ident: Ident, override val span: Span) : Expr(span)

  data class SuperMethodAccess(val ident: Ident, val args: List<Expr>, override val span: Span) :
    Expr(span) {
    init {
      for (it in args) it.parent = this
    }
  }

  data class SuperSubscript(val arg: Expr, override val span: Span) : Expr(span) {
    init {
      arg.parent = this
    }
  }

  data class Subscript(val expr: Expr, val arg: Expr, override val span: Span) : Expr(span) {
    init {
      expr.parent = this
      arg.parent = this
    }
  }

  data class If(val cond: Expr, val then: Expr, val els: Expr, override val span: Span) :
    Expr(span) {
    init {
      cond.parent = this
      then.parent = this
      els.parent = this
    }
  }

  data class Let(val param: Parameter, val binding: Expr, val expr: Expr, override val span: Span) :
    Expr(span) {
    init {
      param.parent = this
      binding.parent = this
      expr.parent = this
    }
  }

  data class FunctionLiteral(val args: List<Parameter>, val expr: Expr, override val span: Span) :
    Expr(span) {
    init {
      expr.parent = this
      for (it in args) it.parent = this
    }
  }

  data class Parenthesised(val expr: Expr, override val span: Span) : Expr(span) {
    init {
      expr.parent = this
    }
  }

  data class New(val type: Type?, val body: ObjectBody, override val span: Span) : Expr(span) {
    init {
      type?.parent = this
      body.parent = this
    }
  }

  data class Amends(val expr: Expr, val body: ObjectBody, override val span: Span) : Expr(span) {
    init {
      expr.parent = this
      body.parent = this
    }
  }

  data class NonNull(val expr: Expr, override val span: Span) : Expr(span) {
    init {
      expr.parent = this
    }
  }

  data class UnaryMinus(val expr: Expr, override val span: Span) : Expr(span) {
    init {
      expr.parent = this
    }
  }

  data class LogicalNot(val expr: Expr, override val span: Span) : Expr(span) {
    init {
      expr.parent = this
    }
  }

  data class BinaryOp(val left: Expr, val right: Expr, val op: Operation, override val span: Span) :
    Expr(span) {
    init {
      left.parent = this
      right.parent = this
    }
  }

  data class TypeTest(val expr: Expr, val type: Type, override val span: Span) : Expr(span) {
    init {
      expr.parent = this
    }
  }

  data class TypeCast(val expr: Expr, val type: Type, override val span: Span) : Expr(span) {
    init {
      expr.parent = this
      type.parent = this
    }
  }
}

enum class Operation {
  POW,
  MULT,
  DIV,
  INT_DIV,
  MOD,
  PLUS,
  MINUS,
  LT,
  GT,
  LTE,
  GTE,
  IS,
  AS,
  EQ_EQ,
  NOT_EQ,
  AND,
  OR,
  PIPE,
  NULL_COALESCE
}
